### PR TITLE
Refactor the ApplicationsView class with MVC in mind

### PR DESCRIPTION
**Pull request by madoar**
This PR refactors the `ApplicationsView` class with MVC in mind. For this purpose a new base component class named `PhoenicisView` has been added.

### CODE REVIEW

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/control/ApplicationView.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import org.phoenicis.javafx.themes.ThemeManager;
+import org.phoenicis.javafx.views.mainwindow.apps.ApplicationFilter;
+import org.phoenicis.repository.dto.ApplicationDTO;
+import org.phoenicis.repository.dto.CategoryDTO;
+import org.phoenicis.scripts.interpreter.ScriptInterpreter;
+
+/**
+ * The component shown inside the Phoenicis "Applications" tab
+ */
+public class ApplicationView extends PhoenicisView<ApplicationView, ApplicationViewSkin> {
```
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/control/ApplicationView.java]: Why did you rename `ApplicationsView` to `ApplicationView`? It shows multiple applications.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/skin/ApplicationViewSkin.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import org.phoenicis.repository.dto.ApplicationDTO;
+import org.phoenicis.repository.dto.CategoryDTO;
+
+import java.util.Comparator;
+import java.util.Optional;
+
+/**
+ * A skin implementation for the {@link ApplicationView} component
+ */
+public class ApplicationViewSkin extends PhoenicisViewSkin<ApplicationView, ApplicationViewSkin> {
```
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/skin/ApplicationViewSkin.java]: `ApplicationViewSkin` vs `ApplicationsViewSkin`

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @param <C> The control type of the component
+ * @param <S> The skin type of the component
+ */
+public abstract class PhoenicisView<C extends PhoenicisView<C, S>, S extends PhoenicisViewSkin<C, S>>
+        extends ControlBase<C, S> {
+    /**
+     * A property containing the initialization state of this view
+     */
+    private final BooleanProperty initialized;
```
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java]: Shouldn't this be `isInitialized`?

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package org.phoenicis.javafx.components.common.control;
+
+import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.SimpleBooleanProperty;
+import org.phoenicis.javafx.components.common.skin.PhoenicisViewSkin;
+
+/**
+ * A base component for a view of the Phoenicis application
```
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java]: I don't really understand the purpose of this class based on the description.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import org.phoenicis.javafx.components.common.control.PhoenicisView;
+import org.phoenicis.javafx.components.common.control.SidebarBase;
+
+/**
+ * A base skin implementation for {@link PhoenicisView} components implementations
+ *
+ * @param <C> The control type of the component
+ * @param <S> The skin type of the component
+ */
+public abstract class PhoenicisViewSkin<C extends PhoenicisView<C, S>, S extends PhoenicisViewSkin<C, S>>
```
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: I don't really like the class name. We should find a better description for what you call `PhoenicisView`. Maybe something like "page" or "tab", not sure currently.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/apps/AppsController.java [hunk 6]
```diff
@@ -19,67 +19,59 @@
...
-    private final RepositoryManager repositoryManager;
-
-    private boolean firstViewSelection = true;
+    private final ApplicationView view;
 
     private Runnable onAppLoaded = () -> {
     };
 
-    public AppsController(ApplicationsView view, RepositoryManager repositoryManager) {
+    public AppsController(ApplicationView view, RepositoryManager repositoryManager) {
```
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/apps/AppsController.java]: Bit inconsistent that it's `AppsController` but `ApplicationView`. We should use:
- `ApplicationsController` and `ApplicationsView`
- `AppsController` and `AppsView`

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java (hunk 3 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java]: I don't think so. The corresponding getter method is named `isInitialized()`.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java (hunk 4 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java]: Ok, I will base the description for the class on the description for `MainWindowView`.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java (hunk 5 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: I don't want to use "tab", because this suggests an implementation of the `Tab` class, which I'm not doing. For "page" I'm missing a direct relation to the usage of the components.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/apps/AppsController.java (hunk 6 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/apps/AppsController.java]: In this PR I want to minimize the number of contained changes which have no direct relation to the refactoring of the `ApplicationsView`. 

Long-term I plan to remove the `AppsController` class. In addition I want to remove the injection mechanism for our main view instances via `@Bean` methods. Instead the creation of the main view instances should be performed inside the `MainWindow` class.

Comment (plata) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/apps/AppsController.java]: Ok.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java (hunk 5 above)
Comment (plata) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: I understand, still I don't like the current naming. I don't really have a different idea though...

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java (hunk 3 above)
Comment (plata) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java]: Fine.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/control/ApplicationView.java (hunk 1 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/control/ApplicationView.java]: Done

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/skin/ApplicationViewSkin.java (hunk 2 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/application/skin/ApplicationViewSkin.java]: Done

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java (hunk 4 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/control/PhoenicisView.java]: Done

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java (hunk 5 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: I agree with you that the name is not ideal, but I think that we can solve this issue later and change the name of the class when we find a better fitting name.
What do you think?

Comment (plata) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: We don't really have another choice unless we find a better name now, have we? Maybe @qparis has an idea.

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: We could also go with `PhoenicisPanel` or `PhoenicisFeature` or `PhoenicisFeaturePanel`.

Comment (plata) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: `FeaturePanel` sounds good to me. Why add `Phoenicis`?

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: Should we also rename our current "views" to `FeaturePanel`, i.e. `ApplicationsView` -> `ApplicationFeaturePanel`? Later on I also want to rename the `MainWindow` class, should we rename it to `PhoenicisApplicationPanel`?

Comment (plata) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: Sounds reasonable.

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/skin/PhoenicisViewSkin.java]: Done

**Comment by plata** [comment]
We should have @qparis approval as well.

**Comment by qparis** [comment]
If you are confortable, I trust you :)

**Comment by madoar** [comment]
A status note:
This PR blocks the remaining refactoring of the JavaFX classes. What is mostly left:
1) refactor the remaining `MainWindowView` classes
2) refactor the `MainWindow` class 
3) remove the controller classes by adding their functionality to the `MainWindowView` classes or to the `MainWindow` class
